docs/proposals/hpa/federated-hpa-v2.md

```diff
@@ -0,0 +1,785 @@
+* When HPA initial assignment, the controller should update the workload's ResourceBinding based on the assignment results.
+* When the member cluster could scale up the workload, FederatedHPA controller should not scale up the workload across clusters, but if the member cluster can not, FederatedHPA controller should scale the workload in other member clusters with different policies.
+* If there is a priority policy, the FederatedHPA controller should scale down the low-priority cluster first, if there is no priority policy, the FederatedHPA controller should not care about the scaling down operation.
+* The HPA resource version will be `autoscaling/v2`, because the new fields introduced in autoscaling/v2 are preserved as annotations when working with autoscaling/v1. [Referring](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+
+To isolate from PP/CPP and avoid PP changes that make it difficult for FederatedHPA to evolve, so we introduce FedertedHPA API to implement the scaling across clusters with rich policies.
+```go
+type FedertedHAP struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    // Spec defines the desired behavior of federated HPA.
+    Spec FederatedHPASpec `json:"spec,omitempty"`
+
+    // Status of the current information about the federated HPA.
+    // +optional
+    Status FederatedHPAStatus `json:"status,omitempty"`
+}
+
+type FederatedHPASpec struct {
+    // ScaleTargetRef points to the target resource to scale, and is used to the pods for which metrics
+    // should be collected, as well as to actually change the replica count.
+    ScaleTargetRef autoscalingv2.CrossVersionObjectReference `json:"scaleTargetRef"`
+
+    // Metrics contains the specifications for which to use to calculate the
+    // desired replica count (the maximum replica count across all metrics will
+    // be used).  The desired replica count is calculated multiplying the
+    // ratio between the target value and the current value by the current
+    // number of pods.  Ergo, metrics used must decrease as the pod count is
+    // increased, and vice-versa.  See the individual metric source types for
+    // more information about how each type of metric must respond.
+    // +optional
+    Metrics []autoscalingv2.MetricSpec `json:"metrics"`
+
+    // behavior configures the scaling behavior of the target
+    // in both Up and Down directions (scaleUp and scaleDown fields respectively).
+    // If not set, the default HPAScalingRules for scale up and scale down are used.
+    // +optional
+    Behavior *autoscalingv2.HorizontalPodAutoscalerBehavior `json:"behavior,omitempty"`
+
+    // ClusterAffinity represents scheduling restrictions to a certain set of clusters.
+    // If not set, any cluster can be scheduling candidate.
+    // +optional
+    ClusterAffinity *policyv1alpha1.ClusterAffinity `json:"clusterAffinity,omitempty"`
```

Comment:
What does this field mean?

Reply:
Same with PP, FederatedHPA should do HPA propagation based on ClusterAffinity(select part of the clusters).